micrometer-core/src/main/java11/io/micrometer/core/instrument/binder/jdk/MicrometerHttpClient.java

```diff
@@ -0,0 +1,292 @@
+            instrumentation.setThrowable(ex);
+            throw ex;
+        }
+        finally {
+            stopObservationOrTimer(instrumentation, request, response);
+        }
+    }
+
+    private <T> void stopObservationOrTimer(
+            ObservationOrTimerCompatibleInstrumentation<HttpClientContext> instrumentation, HttpRequest request,
+            HttpResponse<T> res) {
+        instrumentation.stop(DefaultHttpClientObservationConvention.INSTANCE.getName(), "Timer for JDK's HttpClient",
+                () -> {
+                    Tags tags = Tags.of(HttpClientDocumentedObservation.LowCardinalityKeys.METHOD.asString(),
+                            request.method(), HttpClientDocumentedObservation.LowCardinalityKeys.URI.asString(),
+                            DefaultHttpClientObservationConvention.INSTANCE.getUriTag(request, res, uriMapper));
+                    if (res != null) {
+                        tags = tags.and(Tag.of(HttpClientDocumentedObservation.LowCardinalityKeys.STATUS.asString(),
+                                String.valueOf(res.statusCode())));
+                    }
+                    return tags;
+                });
+    }
+
+    private ObservationOrTimerCompatibleInstrumentation<HttpClientContext> observationOrTimer(
+            HttpRequest.Builder httpRequestBuilder) {
+        return ObservationOrTimerCompatibleInstrumentation.start(this.meterRegistry, this.observationRegistry, () -> {
+            HttpClientContext context = new HttpClientContext(this.uriMapper);
+            context.setCarrier(httpRequestBuilder);
+            return context;
+        }, this.customObservationConvention, DefaultHttpClientObservationConvention.INSTANCE);
+    }
+
+    @Override
+    public <T> CompletableFuture<HttpResponse<T>> sendAsync(HttpRequest httpRequest,
+            HttpResponse.BodyHandler<T> bodyHandler) {
+        return sendAsync(httpRequest, bodyHandler, null);
+    }
+
+    @Override
+    public <T> CompletableFuture<HttpResponse<T>> sendAsync(HttpRequest httpRequest,
+            HttpResponse.BodyHandler<T> bodyHandler, HttpResponse.PushPromiseHandler<T> pushPromiseHandler) {
+        HttpRequest.Builder httpRequestBuilder = decorate(httpRequest);
+        ObservationOrTimerCompatibleInstrumentation<HttpClientContext> instrumentation = observationOrTimer(
+                httpRequestBuilder);
+        HttpRequest request = httpRequestBuilder.build();
+        return client.sendAsync(request, bodyHandler, pushPromiseHandler).handle((response, throwable) -> {
+            if (throwable != null) {
+                instrumentation.setThrowable(throwable);
+            }
+            instrumentation.setResponse(response);
+            stopObservationOrTimer(instrumentation, request, response);
+            return response;
+        });
+    }
+
+    private HttpRequest.Builder decorate(HttpRequest httpRequest) {
+        HttpRequest.Builder builder = HttpRequest.newBuilder(httpRequest.uri());
```

Comment:
We don't need to do decoration if we're not instrumenting with Observation. It would probably be good to have a shortcut to do nothing in that case, as there will be some measurable cost to recreating the builder.

Reply:
This is complicated to implement cleanly. I'm punting on it for now.